target/riscv: added translation drivers

fkhaidari opened the pull request:
Existing flags: 'enable_virtual' and 'enable_virt2phys' were replaced with explicit translation drivers. Motivation:

(1) Having 'enable_virtual' and 'enable_virt2phys' flags set simultaneously may cause double address translation which is unacceptable

(2) Flags were global for all targets which is wrong too

aap-sc commented:
@MarekVCodasip , @JanMatCodasip could you take a look, please?

en-sc reviewed (dismissed):
LGTM (reviewed internally).

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -5093,6 +5098,38 @@ COMMAND_HANDLER(handle_reserve_trigger)
+	if (CMD_ARGC != 1) {
+		command_print(CMD, "Command takes 0 or 1 parameters");
+		return ERROR_COMMAND_ARGUMENT_INVALID;
+	}
```
Detail:

IIRC, the upstream reviewers prefer to return different error code in this case, and to not display any error message. That's because OpenOCD will show the command's help automatically when "SYNTAX_ERROR" is returned.

```suggestion
	if (CMD_ARGC != 1)
		return ERROR_COMMAND_SYNTAX_ERROR;
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -5353,6 +5373,16 @@ static const struct command_registration riscv_exec_command_handlers[] = {
+		.name = "mmu_translation_driver",
+		.handler = handle_riscv_mmu_translation_driver,
+		.mode = COMMAND_ANY,
+		.usage = "'sw'|'hw'|'off'",
```
```suggestion
		.usage = "['sw'|'hw'|'off']",
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
-When on (default), memory accesses are performed on physical or virtual memory
-depending on the current satp configuration. When off, all memory accessses are
-performed on physical memory.
+@deffn {Command} {riscv mmu_translation_driver} [@option{hw}|@option{sw}|@option{off}]
```
I was thinking about a more descriptive name for the command. Does `riscv virt2phys_translation` or `riscv virt2phys_mode` sound good to you?

I'd prefer to not use the word "driver" and instead call it "mode".

```suggestion
@deffn {Command} {riscv virt2phys_translation} [@option{hw}|@option{sw}|@option{off}]
```

or:

```suggestion
@deffn {Command} {riscv virt2phys_mode} [@option{hw}|@option{sw}|@option{off}]
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
-depending on the current satp configuration. When off, all memory accessses are
-performed on physical memory.
+@deffn {Command} {riscv mmu_translation_driver} [@option{hw}|@option{sw}|@option{off}]
+Configure address translation driver:
```
Some reword suggestions for clarity:

```suggestion
Configure how OpenOCD translates virtual addresses to physical:
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
+@itemize @bullet
+@item @option{sw} - virtual address is translated explicitly by OpenOCD using
+traversal of page table entries (by performing physical memory accesses to
+read the respected entries).
```
Reword suggestions and corrections, cont'd:

```suggestion
@item @option{sw} - OpenOCD translates virtual addresses explicitly by
traversing the page table entries (by performing physical memory accesses to
read the respective entries). This is the default mode.
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
+read the respected entries).
+@item @option{hw} - virtual address is translated implicitly by hardware
+(virtual memory access fails with error if hardware doesn't
+support the necessary functionality).
```
```suggestion
@item @option{hw} - Virtual addresses are translated implicitly by hardware.
(Virtual memory access will fail with an error if the hardware doesn't
support the necessary functionality.)
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
+@item @option{hw} - virtual address is translated implicitly by hardware
+(virtual memory access fails with error if hardware doesn't
+support the necessary functionality).
+@item @option{off} - virtual address is not translated (assume identity mapping).
```
```suggestion
@item @option{off} - Virtual addresses are not translated (identity mapping is assumed).
```

fkhaidari replied:
Addressed

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -11356,16 +11356,18 @@ This command can be used to change the memory access methods if the default
+support the necessary functionality).
+@item @option{off} - virtual address is not translated (assume identity mapping).
+@end itemize
+Returns current translation driver if called without arguments.
```
```suggestion
Returns current translation mode if called without arguments.
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -139,6 +139,35 @@ struct tdata1_cache {
 	struct list_head elem_tdata1;
 };
 
+bool riscv_target_config_allows_hw_translation(const struct target *target)
```
Please, consider shortening the name of the function.

You can for example call it `riscv_is_virt2phys_hw_mode()` which is shorter, readable and still very clear.

```suggestion
bool riscv_is_virt2phys_hw_mode(const struct target *target)
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -139,6 +139,35 @@ struct tdata1_cache {
+	return r->translation_driver == RISCV_TRANSLATION_DRIVER_HW;
+}
+
+bool riscv_target_config_allows_sw_translation(const struct target *target)
```
(same as above)

```suggestion
bool riscv_is_virt2phys_sw_mode(const struct target *target)
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -139,6 +139,35 @@ struct tdata1_cache {
+	return r->translation_driver == RISCV_TRANSLATION_DRIVER_SW;
+}
+
+const char *riscv_translation_driver_to_str(riscv_translation_driver_t driver)
```
```suggestion
const char *riscv_virt2phys_mode_to_str(riscv_translation_driver_t driver)
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.h:
```diff
@@ -57,6 +57,14 @@ typedef enum riscv_mem_access_method {
 	RISCV_MEM_ACCESS_MAX_METHODS_NUM
 } riscv_mem_access_method_t;
 
+typedef enum riscv_translation_driver {
+	RISCV_TRANSLATION_DRIVER_HW,
+	RISCV_TRANSLATION_DRIVER_SW,
+	RISCV_TRANSLATION_DRIVER_OFF
+} riscv_translation_driver_t;
```
Similarly to the other comments, this can be renamed. For example:

```suggestion
typedef enum riscv_virt2phys_mode {
	RISCV_VIRT2PHYS_MODE_HW,
	RISCV_VIRT2PHYS_MODE_SW,
	RISCV_VIRT2PHYS_MODE_OFF
} riscv_virt2phys_mode_t;
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv.h:
```diff
@@ -165,6 +173,9 @@ struct riscv_info {
 	 * most recent halt was not caused by a trigger, then this is -1. */
 	int64_t trigger_hit;
 
+	/* The way to translate virtual address to physical */
+	riscv_translation_driver_t translation_driver;
```
```suggestion
	/* The configured approach to translate virtual addresses to physical */
	riscv_virt2phys_mode_t virt2phys_mode;
```

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/startup.tcl:
```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
```
Just a comment - no change requested:

I have not reviewed the changes in this file in detail - I trust you on those.

The new type of tests that @en-sc has proposed [here](https://review.openocd.org/c/openocd/+/8507) would be very good for testing changes like these. Perhaps you are using such tests already internally?

EccoTheDolphin replied:
...

aap-sc replied:
Answered from the wrong account, sorry for spam.

> The new type of tests that @en-sc has proposed [here](https://review.openocd.org/c/openocd/+/8507) would be very good for testing changes like these. Perhaps you are using such tests already internally?

1. These types of tests would be useful, though integrating it to that infrastructure will be difficult, since these specific tests require a specific response from the target (dummy adapter just  guarantees that some results are returned, not a valid one). 
2. This commit is tested internally against using internal testsuite that I'm still in the process of opensourcing :(. Those test just generate MMU pages on the fly, flip the switch for translation mode and try to do memory reads/writes  comparing results with the expected one.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -5353,6 +5373,16 @@ static const struct command_registration riscv_exec_command_handlers[] = {
+		.handler = handle_riscv_mmu_translation_driver,
+		.mode = COMMAND_ANY,
+		.usage = "'sw'|'hw'|'off'",
+		.help = "Configure address translation way: "
```
```suggestion
		.help = "Configure the virtual address translation mode: "
```

or

```suggestion
		.help = "Configure the virtual address translation mechanism: "
```

or similar

fkhaidari replied:
Addressed

JanMatCodasip commented on src/target/riscv/riscv-013.c:
```diff
@@ -1974,7 +1974,7 @@ static int examine(struct target *target)
 
-	if (info->progbufsize < 4 && riscv_enable_virtual) {
+	if (info->progbufsize < 4 && riscv_target_config_allows_hw_translation(target)) {
 		LOG_TARGET_ERROR(target, "set_enable_virtual is not available on this target. It "
```
The old command name (`set_enable_virtual`) in the message should be replaced by the new name.

JanMatCodasip commented on src/target/riscv/riscv-013.c:
```diff
@@ -1974,7 +1974,7 @@ static int examine(struct target *target)
 				info->impebreak);
 	}
 
-	if (info->progbufsize < 4 && riscv_enable_virtual) {
```
The program buffer size check `info->progbufsize < 4` does not look right because it does not account for `impebreak`. Can you please check it.

I believe `has_sufficient_progbuf()` should be used instead.

fkhaidari replied:
I will address this in future MR

JanMatCodasip replied:
Ok, thank you.

JanMatCodasip reviewed (commented):
This change looks very good to me and makes the TCL interface for virt2phys configuration much clearer than it is now.

I have only minor comments, mostly just suggestions for renaming or reword in documentation. Please if you can check them out.

fk-sc commented:
@JanMatCodasip thank you for the review. All threads were addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -5353,6 +5371,16 @@ static const struct command_registration riscv_exec_command_handlers[] = {
+		.usage = "['sw'|'hw'|'off']",
+		.help = "Configure the virtual address translation mode: "
+				"sw - translate vaddr to paddr by manually traversing page tables, "
+				"hw - translate vaddr to paddr by hardware,"
```
Really a detail: Added trailing space character:

```suggestion
				"hw - translate vaddr to paddr by hardware, "
```

fkhaidari replied:
Addressed

JanMatCodasip reviewed (dismissed):
LGTM, thank you.

fkhaidari committed:
target/riscv: added translation drivers

Existing flags: 'enable_virtual' and 'enable_virt2phys' were
replaced with explicit translation drivers. Motivation:

(1) Having 'enable_virtual' and 'enable_virt2phys' flags set simultaneously
may cause double address translation which is unacceptable

(2) Flags were global for all targets which is wrong too

Signed-off-by: Farid Khaydari <f.khaydari@syntacore.com>
